hubblestack/hec/dq.py

```diff
@@ -1,129 +1,376 @@
-        return s
+        # Is it "dangerous" to unlink files during the os.walk (via generator)?
+        # .oO( probably doesn't matter )
+        r = b''
+        for fname in self.files:
+            with open(fname, 'rb') as fh:
+                p = self.decompress(fh.read())
+            if len(r) + len(self.sep) + len(p) > sz:
+                break
+            if r:
+                r += self.sep
+            r += p
+            os.unlink(fname)
+        self._count()
+        return r
 
-    def _drop_for_items(self):
-        if self.max_items and self.max_items > 0:
-            log.debug('_drop_for_size .max_items %d', self.max_items)
-            while self.eventcount > self.max_items:
-                self.pop()
-
-    def _drop_for_size(self):
-        if self.max_size and self.max_size >= 1000:
-            log.debug('_drop_for_size .max_size %d', self.max_size)
-            while self.disksize > self.max_size and self.eventcount > 0:
-                self.pop()
-
-    def push(self, item):
-        log.debug('pushing item to diskqueue')
-        if self.restrict_to and not isinstance(item, self.restrict_to):
-            raise DataFormatError("items in this queue must be of type {0}".format(self.restrict_to))
-        q = self.q
-        q.append(item)
-        self._drop_for_items()
-        self._drop_for_size()
-        self.last_put = time.time()
-    put = push
-
-    def peek(self, idx=0):
-        try:
-            return self.q[idx]
-        except IndexError:
-            pass
-
-    def pull(self):
-        if self.eventcount < 1:
-            log.debug('pull: empty')
-            return
+    def pop(self):
+        ''' remove the next item from the queue (do not return it); useful with .peek() '''
+        for fname in self.files:
+            os.unlink(fname)
+            break
+        self._count()
+
+    @property
+    def files(self):
+        ''' generate all filenames in the diskqueue (returns iterable) '''
+        def _k(x):
+            try:
+                return [ int(i) for i in x.split('.') ]
+            except:
+                pass
+            return x
+        for path, dirs, files in sorted(os.walk(self.directory)):
+            for fname in [ os.path.join(path, f) for f in sorted(files, key=_k) ]:
+                yield fname
+
+    def _count(self):
+        self.cn = 0
+        self.sz = 0
+        for fname in self.files:
+            self.sz += os.stat(fname).st_size
+            self.cn += 1
+        log.debug('disk cache sizes: cn=%d sz=%d', self.cn, self.sz)
+
+    @property
+    def msz(self):
+        ''' The size of the queue as it would be returned from getz() iff getz had no size limit '''
+        return self.sz + max(0, len(self.sep) * (self.cn -1))
+
+    def __len__(self):
+        return self.msz
+
+class DiskBackedQueue:
+    def __init__(self, directory, mem_size=DEFAULT_MEMORY_SIZE,
+                            disk_size=DEFAULT_DISK_SIZE, ok_types=OK_TYPES, fresh=False):
```

Comment:
I think the new indentation must have been an accident, or else you're trolling me.

```
class DiskBackedQueue:
        def __init__(self, directory, mem_size=DEFAULT_MEMORY_SIZE,
                disk_size=DEFAULT_DISK_SIZE, ok_types=OK_TYPES, fresh=False):
```
is what we're looking for.